packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js

```diff
@@ -64,32 +68,188 @@ function Modal({ isOpen, onClose }) {
+      );
+    }
+
+    const audioDeviceExists =
+      audioInput &&
+      mediaDevices.some((device) => device.deviceId === audioInput);
+
+    // Audio device not set yet, or was detached. Choose first available audio device.
+    if (!audioInput || !audioDeviceExists) {
+      setAudioInput(
+        mediaDevices.find((device) => device.kind === 'audioinput')?.deviceId
+      );
+    }
+  }, [audioInput, mediaDevices, videoInput]);
+
+  useEffect(() => {
+    navigator.mediaDevices.addEventListener('devicechange', updateMediaDevices);
+
+    return () => {
+      navigator.mediaDevices.removeEventListener(
+        'devicechange',
+        updateMediaDevices
+      );
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- We only want to run it once.
+  }, []);
+
+  const videoInputs = mediaDevices
+    .filter((device) => device.kind === 'videoinput')
+    .map(({ deviceId, label }) => ({ value: deviceId, label }));
+  const audioInputs = mediaDevices
+    .filter((device) => device.kind === 'audioinput')
+    .map(({ deviceId, label }) => ({ value: deviceId, label }));
 
   const onStop = useCallback((blob) => {
     const f = blobToFile(blob, 'test-recording.webm', 'video/webm');
     setFile(f);
   }, []);
 
-  const uploadWithPreview = useUploadWithPreview();
+  const onError = (err) => {
+    // eslint-disable-next-line no-console -- TODO: Telemetry, notify user - UX pending.
+    console.log(err);
+  };
+
   const {
     error,
     status,
     mediaBlob,
     stopRecording,
     startRecording,
     liveStream,
+    getMediaStream,
     clearMediaStream,
     clearMediaBlob,
   } = useMediaRecorder({
     recordScreen: false,
     blobOptions: { type: 'video/webm' },
     mediaStreamConstraints: {
-      audio: enableAudio,
-      video: enableVideo,
+      audio: audioInput && enableAudio ? { deviceId: audioInput } : enableAudio,
+      video: videoInput && enableVideo ? { deviceId: videoInput } : enableVideo,
     },
-    onStop: onStop,
+    onStop,
+    onError,
   });
 
+  // eslint-disable-next-line react-hooks/exhaustive-deps -- Only run once.
+  useEffect(() => clearMediaStream, []);
+
+  const previouslyHadStream = usePrevious(Boolean(liveStream));
+  const hasStream = Boolean(liveStream);
+  const wasOpen = usePrevious(isOpen);
+  const isRecording = status === 'recording';
+  const wasRecording = usePrevious(isRecording);
+
+  // Ensure the stream is refreshed after switching video or audio input
+  // (which clears the stream).
+  useEffect(() => {
+    if (
+      previouslyHadStream &&
+      !hasStream &&
+      !isImageCapture &&
+      !mediaBlob &&
+      !isRecording &&
+      !wasRecording
+    ) {
+      getMediaStream();
+    }
+  }, [
+    previouslyHadStream,
+    hasStream,
+    getMediaStream,
+    isImageCapture,
+    mediaBlob,
+    isRecording,
+    wasRecording,
+  ]);
+
+  // Try to get permissions as soon as the modal opens.
+  useEffect(() => {
```

Comment:
This should help with the flow  👍 even for testing the permissions have been a problem.